frontend/controller/service-worker.js

```diff
@@ -92,128 +94,51 @@ sbp('sbp/selectors/register', {
-              payload: JSON.stringify(subscription.toJSON())
+          }
+          return subscription
+        }).catch(e => {
+          if (!(retryCount > 3) && e?.message === 'WebSocket connection is not open') {
+            sbp('okTurtles.events/once', ONLINE, () => {
+              setTimeout(() => sbp('service-worker/setup-push-subscription', (retryCount || 0) + 1), 200)
             })
-
-            resolve()
-          } catch (err) {
-            console.error('[sw] service-worker/setup-push-subscription failed with the following error: ', err)
-            resolve()
+            return
           }
+          console.error('[service-worker/setup-push-subscription] Error setting up push subscription', e)
+          alert(e?.message || 'Error')
         })
+        : null
 
-        messageToPushServerIfSocketConnected({ action: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY })
+      await sbp('push/reportExistingSubscription', existingSubscription?.toJSON()).catch(e => {
+        console.error('[service-worker/setup-push-subscription] Error reporting existing subscription', e)
       })
-    }
-  },
-  'service-worker/send-push': async function (payload) {
-    if (!('serviceWorker' in navigator) || !('Notification' in window)) { return }
-
-    if (Notification.permission !== 'granted') {
-      console.debug('[sw] stopped sending a push-notification data to the server because of the permission not granted.')
-      return
-    }
-
-    const swRegistration = await navigator.serviceWorker.ready
-
-    if (!swRegistration) {
-      console.error('No service-worker registration found!')
-      return
-    }
-
-    const pushSubscription = await swRegistration.pushManager.getSubscription()
-    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
-
-    if (pushSubscription) {
-      pubsub.socket.send(createMessage(
-        REQUEST_TYPE.PUSH_ACTION,
-        {
-          action: PUSH_SERVER_ACTION_TYPE.SEND_PUSH_NOTIFICATION,
-          payload: JSON.stringify({ ...payload, endpoint: pushSubscription.endpoint })
-        }
-      ))
-    } else {
-      console.error('No existing push subscription found!')
-    }
-  },
-  'service-worker/check-push-subscription-ready': async function () {
-    if (!('serviceWorker' in navigator) || !('Notification' in window)) { return false }
-
-    const swRegistration = await navigator.serviceWorker.ready
-    if (swRegistration) {
-      const pushSubscription = await swRegistration.pushManager.getSubscription()
-
-      return !!pushSubscription
-    } else { return false }
-  },
-  'service-worker/resubscribe-push': function (subscription) {
-    const pubsub = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
-
-    pubsub.socket.send(createMessage(
-      REQUEST_TYPE.PUSH_ACTION,
-      {
-        action: PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION,
-        payload: JSON.stringify(subscription.toJSON())
-      }
-    ))
+      return true
```

Comment:
So I'm a bit confused by the overall organization and lifecycle that's going on here, and I think it deserves at least a comment (right above the definition of `'service-worker/setup-push-subscription'`).

I noticed that `'push/reportExistingSubscription'` is called twice in `push.js` - which is running in the service worker. Once on reconnection, and once on `pushsubscriptionchange` event.

However, it's also called here, *outside of the service worker*, in this ``'service-worker/setup-push-subscription'`` selector.

So my question is: why? Why does ``'service-worker/setup-push-subscription'`` exist outside of the service worker? Why isn't it called inside of the service worker at start, and maybe whenever it's awoken?

Could this be properly elaborated on and explained in a comment above ``'service-worker/setup-push-subscription'``?

Reply:
Most or all of the examples of SW do this setup _outside_ of the SW (as is being done here). When writing the demo notifications app, I tried putting this logic in the SW, but it didn't seem to work as reliably.

With that said:

> I noticed that 'push/reportExistingSubscription' is called twice in push.js - which is running in the service worker. Once on reconnection, and once on pushsubscriptionchange event.
> However, it's also called here, outside of the service worker, in this 'service-worker/setup-push-subscription' selector.

This is because there are three reasons to report the existing subscription. On reconnection covers existing subscriptions, the subscription change is self-explanatory and the third reason is for a _new_ subscription. This is the case here.

>  Why isn't it called inside of the service worker at start, and maybe whenever it's awoken?

Because that'd miss new subscriptions. Although in theory creating the subscription can be done in the SW, this doesn't work as reliably, and, importantly, part of this logic **needs** to be outside of the SW because requesting the notification permission can only be done outside of the SW.

> Could this be properly elaborated on and explained in a comment above 'service-worker/setup-push-subscription'?

Sure